<html>
<head>
<title>error.py</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
.s0 { color: #a71d5d;}
.s1 { color: #333333;}
.s2 { color: #63a35c;}
.s3 { color: #969896;}
.s4 { color: #0086b3;}
.s5 { color: #183691;}
.ln { color: #333333; font-weight: normal; font-style: normal; }
</style>
</head>
<body bgcolor="#ffffff">
<table CELLSPACING=0 CELLPADDING=5 COLS=1 WIDTH="100%" BGCOLOR="#c0c0c0" >
<tr><td><center>
<font face="Arial, Helvetica" color="#000000">
error.py</font>
</center></td></tr></table>
<pre><a name="l1"><span class="ln">1    </span></a><span class="s0">from </span><span class="s1">enum </span><span class="s0">import </span><span class="s1">Enum</span>
<a name="l2"><span class="ln">2    </span></a>
<a name="l3"><span class="ln">3    </span></a>
<a name="l4"><span class="ln">4    </span></a><span class="s0">class </span><span class="s1">ExportErrorType</span><span class="s2">(</span><span class="s1">Enum</span><span class="s2">)</span><span class="s0">:</span>
<a name="l5"><span class="ln">5    </span></a>    <span class="s3"># User providing invalid inputs to either tracer, or other public facing APIs</span>
<a name="l6"><span class="ln">6    </span></a>    <span class="s1">INVALID_INPUT_TYPE </span><span class="s0">= </span><span class="s4">1</span>
<a name="l7"><span class="ln">7    </span></a>
<a name="l8"><span class="ln">8    </span></a>    <span class="s3"># User returning values from their models that we don't support.</span>
<a name="l9"><span class="ln">9    </span></a>    <span class="s1">INVALID_OUTPUT_TYPE </span><span class="s0">= </span><span class="s4">2</span>
<a name="l10"><span class="ln">10   </span></a>
<a name="l11"><span class="ln">11   </span></a>    <span class="s3"># Generated IR does not conform to Export IR Specification.</span>
<a name="l12"><span class="ln">12   </span></a>    <span class="s1">VIOLATION_OF_SPEC </span><span class="s0">= </span><span class="s4">3</span>
<a name="l13"><span class="ln">13   </span></a>
<a name="l14"><span class="ln">14   </span></a>    <span class="s3"># User's code contains types and functionalities we don't support.</span>
<a name="l15"><span class="ln">15   </span></a>    <span class="s1">NOT_SUPPORTED </span><span class="s0">= </span><span class="s4">4</span>
<a name="l16"><span class="ln">16   </span></a>
<a name="l17"><span class="ln">17   </span></a>    <span class="s3"># User's code didn't provide necessary details for us to successfully trace and export.</span>
<a name="l18"><span class="ln">18   </span></a>    <span class="s3"># For example, we use a lot of decorators and ask users to annotate their model.</span>
<a name="l19"><span class="ln">19   </span></a>    <span class="s1">MISSING_PROPERTY </span><span class="s0">= </span><span class="s4">5</span>
<a name="l20"><span class="ln">20   </span></a>
<a name="l21"><span class="ln">21   </span></a>    <span class="s3"># User is using an API without proper initialization step.</span>
<a name="l22"><span class="ln">22   </span></a>    <span class="s1">UNINITIALIZED </span><span class="s0">= </span><span class="s4">6</span>
<a name="l23"><span class="ln">23   </span></a>
<a name="l24"><span class="ln">24   </span></a>
<a name="l25"><span class="ln">25   </span></a><span class="s0">def </span><span class="s1">internal_assert</span><span class="s2">(</span><span class="s1">pred</span><span class="s0">: </span><span class="s1">bool</span><span class="s2">, </span><span class="s1">assert_msg</span><span class="s0">: </span><span class="s1">str</span><span class="s2">) </span><span class="s1">-&gt; </span><span class="s0">None:</span>
<a name="l26"><span class="ln">26   </span></a>    <span class="s3">&quot;&quot;&quot; 
<a name="l27"><span class="ln">27   </span></a>    This is exir's custom assert method. It internally just throws InternalError. 
<a name="l28"><span class="ln">28   </span></a>    Note that the sole purpose is to throw our own error while maintaining similar syntax 
<a name="l29"><span class="ln">29   </span></a>    as python assert. 
<a name="l30"><span class="ln">30   </span></a>    &quot;&quot;&quot;</span>
<a name="l31"><span class="ln">31   </span></a>
<a name="l32"><span class="ln">32   </span></a>    <span class="s0">if not </span><span class="s1">pred</span><span class="s0">:</span>
<a name="l33"><span class="ln">33   </span></a>        <span class="s0">raise </span><span class="s1">InternalError</span><span class="s2">(</span><span class="s1">assert_msg</span><span class="s2">)</span>
<a name="l34"><span class="ln">34   </span></a>
<a name="l35"><span class="ln">35   </span></a>
<a name="l36"><span class="ln">36   </span></a><span class="s0">class </span><span class="s1">InternalError</span><span class="s2">(</span><span class="s1">Exception</span><span class="s2">)</span><span class="s0">:</span>
<a name="l37"><span class="ln">37   </span></a>    <span class="s3">&quot;&quot;&quot; 
<a name="l38"><span class="ln">38   </span></a>    Raised when an internal invariance is violated in EXIR stack. 
<a name="l39"><span class="ln">39   </span></a>    Should hint users to report a bug to dev and expose the original 
<a name="l40"><span class="ln">40   </span></a>    error message. 
<a name="l41"><span class="ln">41   </span></a>    &quot;&quot;&quot;</span>
<a name="l42"><span class="ln">42   </span></a>
<a name="l43"><span class="ln">43   </span></a>    <span class="s0">def </span><span class="s1">__init__</span><span class="s2">(</span><span class="s1">self</span><span class="s2">, </span><span class="s1">message</span><span class="s0">: </span><span class="s1">str</span><span class="s2">) </span><span class="s1">-&gt; </span><span class="s0">None:</span>
<a name="l44"><span class="ln">44   </span></a>        <span class="s1">super</span><span class="s2">().</span><span class="s1">__init__</span><span class="s2">(</span><span class="s1">message</span><span class="s2">)</span>
<a name="l45"><span class="ln">45   </span></a>
<a name="l46"><span class="ln">46   </span></a>
<a name="l47"><span class="ln">47   </span></a><span class="s0">class </span><span class="s1">ExportError</span><span class="s2">(</span><span class="s1">Exception</span><span class="s2">)</span><span class="s0">:</span>
<a name="l48"><span class="ln">48   </span></a>    <span class="s3">&quot;&quot;&quot; 
<a name="l49"><span class="ln">49   </span></a>    This type of exception is raised for errors that are directly caused by the user 
<a name="l50"><span class="ln">50   </span></a>    code. In general, user errors happen during model authoring, tracing, using our public 
<a name="l51"><span class="ln">51   </span></a>    facing APIs, and writing graph passes. 
<a name="l52"><span class="ln">52   </span></a>    &quot;&quot;&quot;</span>
<a name="l53"><span class="ln">53   </span></a>
<a name="l54"><span class="ln">54   </span></a>    <span class="s0">def </span><span class="s1">__init__</span><span class="s2">(</span><span class="s1">self</span><span class="s2">, </span><span class="s1">error_code</span><span class="s0">: </span><span class="s1">ExportErrorType</span><span class="s2">, </span><span class="s1">message</span><span class="s0">: </span><span class="s1">str</span><span class="s2">) </span><span class="s1">-&gt; </span><span class="s0">None:</span>
<a name="l55"><span class="ln">55   </span></a>        <span class="s1">prefix </span><span class="s0">= </span><span class="s5">f&quot;[{</span><span class="s1">error_code</span><span class="s5">}]: &quot;</span>
<a name="l56"><span class="ln">56   </span></a>        <span class="s1">super</span><span class="s2">().</span><span class="s1">__init__</span><span class="s2">(</span><span class="s1">prefix </span><span class="s0">+ </span><span class="s1">message</span><span class="s2">)</span>
<a name="l57"><span class="ln">57   </span></a></pre>
</body>
</html>